File: Problem-1.cpp
// https://leetcode.com/problems/subarray-sum-equals-k/

// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
// use a hashmap that stores the sum of elements at each index and the frequency.
// Initialize the hashmap with key = 0, val = 1, for the case when the first element of the array is equal to k.
// Take the sum of the element at each index, check if the differece( sum - k) is already present, if yes then increment the count. 


class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int,int> mp;
        mp[0] = 1;
        int count=0; int sum = 0;
        for(int i=0; i<nums.size(); i++){
            sum += nums[i];
            int diff = sum - k;
            count += mp[diff];
            mp[sum] += 1;
        }
        return count;
    }
};

File: Problem_1.js
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
// Your code here along with comments explaining your approach: Using running sum method as taught in class
/**
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
    let rSum = 0;
    let rSumMap = new Map();
    rSumMap.set(0, 1);
    let subArrCount = 0;

    for(let i=0; i<nums.length; i++){
        rSum += nums[i];
        if(rSumMap.has(rSum - k)){
            subArrCount+= rSumMap.get(rSum - k);
        }
        if(rSumMap.has(rSum)){
            let tmp = rSumMap.get(rSum);
            rSumMap.set(rSum, tmp+1);
        }
        else
            rSumMap.set(rSum, 1);   
        
    }
    return subArrCount;
};